transport/udp/udp.go

```diff
@@ -0,0 +1,390 @@
+}
+
+func (uf *udpFullPipe) Recv() (*mangos.Message, error) {
+	n, err := uf.UDPConn.Read(uf.readBuf)
+	if err != nil {
+		return nil, err
+	}
+	msg, err := uf.extractMsgFromPacket(uf.readBuf[:n])
+	if err != nil {
+		return nil, err
+	}
+	return msg, err
+}
+
+func (uf *udpFullPipe) Send(m *mangos.Message) error {
+	return uf.sendHelper(m, uf.UDPConn.Write)
+}
+
+// udpHalfPipe is a half-open connection associated with a remote peer, created by the udp PipeListener.
+type udpHalfPipe struct {
+	*udpPipe
+	peer    net.Addr             // the remote endpoint talking to this Listener
+	inqueue chan *mangos.Message // buffers incoming messages until Recv() is called (CAVEAT: udpMaxBacklog)
+}
+
+func (uh udpHalfPipe) String() string {
+	return fmt.Sprintf("udp half pipe: %s -> %s", uh.LocalAddr(), uh.RemoteAddr())
+}
+
+// RemoteAddr overrides the function of the same name provided by UDPConn, since for listening endpoints
+// it always returns nil. The @peer value is non-nil once a remote endpoint has connected to the listener.
+func (uh *udpHalfPipe) RemoteAddr() net.Addr {
+	return uh.peer
+}
+
+func (uh *udpHalfPipe) Recv() (*mangos.Message, error) {
+	if msg, ok := <-uh.inqueue; ok {
+		return msg, nil
+	}
+	return nil, mangos.ErrClosed
+}
+
+// Send mimics writing to a connected peer via a udp half-pipe.
+func (uh *udpHalfPipe) Send(m *mangos.Message) error {
+	return uh.sendHelper(m, func(b []byte) (int, error) {
+		return uh.UDPConn.WriteTo(b, uh.peer)
+	})
+}
+
+func (uh *udpHalfPipe) Close() error {
+	halfPipePeers.Lock()
+	delete(halfPipePeers.byAddr, uh.peer.String())
+	close(uh.inqueue)
+	halfPipePeers.Unlock()
+
+	return uh.udpPipe.Close()
+}
+
+/*
+ * Dialer Methods
+ */
+
+// dialer inherits from udpEndpoint
+type dialer struct {
+	*udpEndpoint
+}
+
+func (d *dialer) Dial() (mangos.Pipe, error) {
+	conn, err := net.DialUDP("udp", nil, d.addr)
+	if err != nil {
+		return nil, err
+	}
+	return &udpFullPipe{
+		udpPipe: &udpPipe{
+			udpEndpoint: udpEndpoint{
+				addr:  d.addr,
+				proto: d.proto,
+				opts:  d.opts,
+			},
+			UDPConn: conn,
+			open:    true,
+			props: map[string]interface{}{
+				mangos.PropLocalAddr:  conn.LocalAddr(),
+				mangos.PropRemoteAddr: conn.RemoteAddr(),
+			},
+		},
+		readBuf: make([]byte, udpMaxBuf),
+	}, nil
+}
+
+/*
+ * Listener methods
+ */
+type listener struct {
+	*udpEndpoint
+	*net.UDPConn // implements listener.Close()
+}
+
+func (l *listener) Address() string {
+	return "udp://" + l.addr.String()
+}
+
+func (l *listener) Listen() (err error) {
+	l.UDPConn, err = net.ListenUDP("udp", l.addr)
+	return
+}
+
+// halfPipePeers records remote endpoints that have connected to this Listener
+var halfPipePeers = struct {
+	// which peer connected the listening endpoint
+	byAddr map[string]*udpHalfPipe
+	sync.Mutex
+}{byAddr: make(map[string]*udpHalfPipe)}
+
+// Accept sorts messages into the mailboxes of existing peers, until a new
+// peer connects; causing a new udpHalfPipe to be returned.
```

Comment:
You've tried to make a pipe/connection oriented interface here.  I think this isn't quite the way I'd go.  UDP is by its very nature disconnected.

I'd instead go with a different implementation, where Listen() does the bind, but Accept() is just a noop to return a suitable reference to the single listening pipe (perhaps wrapped in another object).

I think this would be simpler than the whole full/half pipe thing.

Under this scenario, UDP applications would never see more than a single "pipe" for a given listening address, and
for a given "connect" address they would accept packets from anywhere (or perhaps the option to discard from "unconnected" peers could be a socket option?)

I want to think on this a bit.